supermarq/features.py

```diff
@@ -0,0 +1,170 @@
+    Liveness feature = sum of all entries in the liveness matrix / (num_qubits * depth).
+
+    Input
+    -----
+    circuit : A cirq or qiskit quantum circuit
+    """
+
+    if isinstance(circuit, cirq.Circuit):
+        circ = sm.converters.cirq_to_qiskit(circuit)
+    else:
+        circ = circuit
+
+    num_qubits = circ.num_qubits
+    dag = qiskit.converters.circuit_to_dag(circ)
+    dag.remove_all_ops_named("barrier")
+
+    activity_matrix = np.zeros((num_qubits, dag.depth()))
+
+    for i, layer in enumerate(dag.layers()):
+        for op in layer["partition"]:
+            for qubit in op:
+                activity_matrix[qubit.index, i] = 1
+
+    return np.sum(activity_matrix) / (num_qubits * dag.depth())
+
+
+def compute_parallelism(circuit: Union[cirq.Circuit, qiskit.circuit.QuantumCircuit]) -> float:
+    """Compute the parallelism of the given quantum circuit.
+
+    Parallelism feature = max(1 - depth / # of gates, 0)
+
+    Input
+    -----
+    circuit : A cirq or qiskit quantum circuit
+    """
+
+    if isinstance(circuit, cirq.Circuit):
+        circ = sm.converters.cirq_to_qiskit(circuit)
+    else:
+        circ = circuit
+
+    dag = qiskit.converters.circuit_to_dag(circ)
+    dag.remove_all_ops_named("barrier")
+    return max(1 - (circ.depth() / len(dag.gate_nodes())), 0)
+
+
+def compute_measurement(circuit: Union[cirq.Circuit, qiskit.circuit.QuantumCircuit]) -> float:
+    """Compute the measurement feature of the given quantum circuit.
+
+    Measurement feature = # of layers of mid-circuit measurement / circuit depth.
+
+    Input
+    -----
+    circuit : A cirq or qiskit quantum circuit
+    """
+    if isinstance(circuit, cirq.Circuit):
+        circ = sm.converters.cirq_to_qiskit(circuit)
+    else:
+        circ = circuit
+
+    circ.remove_final_measurements()
+    dag = qiskit.converters.circuit_to_dag(circ)
+    dag.remove_all_ops_named("barrier")
+
+    reset_moments = 0
+    gate_depth = dag.depth()
+
+    for layer in dag.layers():
+        reset_present = False
+        for op in layer["graph"].op_nodes():
+            if op.name == "reset":
+                reset_present = True
+        if reset_present:
+            reset_moments += 1
+
+    return reset_moments / gate_depth
+
+
+def compute_entanglement(circuit: Union[cirq.Circuit, qiskit.circuit.QuantumCircuit]) -> float:
+    """Compute the entanglement-ratio of the given quantum circuit.
+
+    Entanglement-ratio = ratio between # of 2-qubit gates and total number of gates in the circuit.
+
+    Input
+    -----
+    circuit : A cirq or qiskit quantum circuit
+    """
+
+    if isinstance(circuit, cirq.Circuit):
+        circ = sm.converters.cirq_to_qiskit(circuit)
+    else:
+        circ = circuit
+
+    dag = qiskit.converters.circuit_to_dag(circ)
+    dag.remove_all_ops_named("barrier")
+
+    return len(dag.two_qubit_ops()) / len(dag.gate_nodes())
+
+
+def compute_depth(circuit: Union[cirq.Circuit, qiskit.circuit.QuantumCircuit]) -> float:
```

Comment:
Get rid of `qiskit.QuantumCircuit`

Reply:
Since the features need `qiskit.QuantumCircuit` anyway, it seems very easy to support both. I agree that we can keep the benchmark generators simple -> only generate cirq circuits. But the features can easily handle both

Reply:
Ah, sure. I see now. This is fine.

Reply:
Actually, still haven't' decided. I think we want people to do `compute_depth(cirq_to_qiskit(qiskit_circuit)`. T

Reply:
If our benchmarks are always returning cirq circuits, it doesn't make sense to me to force them to add this extra code when they want to compute the feature vectors. They should be able to enter in the exact output of the benchmark without needing to alter it

Reply:
If our benchmarks are always returning cirq circuits, it doesn't make sense to me to force them to add this extra code when they want to compute the feature vectors. They should be able to enter in the exact output of the benchmark without needing to alter it